test/library/standard/Sort/correctness/sort-array-of-owned.chpl

```diff
@@ -8,14 +8,16 @@ class Value {
 }
 
 
-record ValueComparator {
-  proc key(v: Value?) where useKeyComparator {
+record keyValueComparator {
```

Comment:
These test changes seems to imply we've changed behavior for whether it's okay to have both a key and a compare method on a type.  I was under the impression we weren't changing what was allowed, am I misremembering?

Reply:
I see this as a test trick. `useKeyComparator` is a param, so only `key` or `compare` can ever be compiled at one time. I can't think of a reason users would write code this way

Reply:
We're supporting all three comparator styles, but I could easily see an extension library not wanting to bother and only supporting one.  And if multiple libraries take that tactic but don't agree on which style of comparator they like more, a user would need to write code like this to work with both of them, right?

Reply:
But today users can't write code like this. You can only have either `key` or `compare`, so in your scenario users need to recompile anyways

Reply:
The param doesn't have to be a config, it could be a field on the type.  The main benefit of doing that would be not having to write and support a completely separate type and type name for what is essentially the same purpose.  And the type being two different instantiations wouldn't be too much of a problem, since the user probably isn't storing multiple comparators at a time in an array, they're probably using it as an argument to functions which can be written to use the generic version of the type

Reply:
So for example, a program like this?

```chapel
use Sort;
record cmp: keyComparator, relativeComparator {
  type elt;
  proc key(a) where elt == int do return a;
  proc compare(a, b) where elt == real do return (a < b):int;

}
var a: [1..10] int;
var b: [1..10] real;
sort(a, new cmp(a.eltType));
sort(b, new cmp(b.eltType));
```

I don't think we can write this pattern with interfaces. Because if `elt` is `int`, then `key` is resolved and we implement `keyComparator`. But, then we have said that `cmp` implements `relativeComparator` but have not defined a `compare`, so the interface is unfulfilled. And vice-versa if `elt` is real.

Reply:
Okay, that makes sense

Reply:
I think there is an opportunity for something like this

```chapel
use Sort;
config type elt;
record cmp {
  proc key(a) where elt == int do return a;
  proc compare(a, b) where elt == real do return (a < b):int;
}
cmp implements (if elt == int then keyComparator else relativeComparator);
var a: [1..10] int;
var b: [1..10] real;
sort(a, new cmp(a.eltType));
sort(b, new cmp(b.eltType));
```

But the parser doesn't accept this. Rewriting it as `cmp implements myTypeFunction()` segfaults (https://github.com/chapel-lang/chapel/issues/25824).

Also rewriting it as an `if` statement instead of expression also doesn't seem to work, the compiler just ignores the `implements`.

So none of these work. Furthermore, even if these patterns compiled they wouldn't work with `type elt` as a field

Reply:
That's a shame, but makes sense.  Thanks for exploring this possibility with me, I think it's okay for us to accept how the PR currently handles it